### PR TITLE
feat(developer): Support loading XML LDML keyboards in TIKE 🦕

### PR DESCRIPTION
**Pull request by mcdurdin**
This PR may be more easily reviewed commit-by-commit.

* removed obsolete project code: Removes old state load/save from each project file type, and the unused WarnAsError property, in preparation for xmlLdmlProjectFile.

* ldmlKeyboard3.dtd

* Add basic LDML XML keyboard file support to TIKE projects

* Split dmActionsDebugger out of dmActionsKeyboardEditor:
  * Because the debugger actions are used only with the kmn keyboard editor, it makes sense to keep them as a separate module, in preparation for supporting XmlLdmlKeyboards.

* add xmlLdml types and form, rename ActiveEditor:
  * Adds XmlLdml keyboard editor form based on the plain text Editor for now.
  * ActiveEditor is a name used for multiple purposes. Rename to ActiveKmnKeyboardEditor and introduce ActiveLdmlKeyboardEditor alongside.

* Actions to support LdmlKeyboardEditor:
  * Adds actions to support LdmlKeyboardEditor, and refactors and reorganizes the existing actions along the way to make them clearer.
  * Compile, Install, and Uninstall supported at this time

@keymanapp-test-bot skip

A full user test of the new XML keyboard functionality will be included in a subsequent PR; there are a number of incomplete features at present which make test difficult.

### CODE REVIEW

**Comment by ermshiperete** [comment]
GHA packaging builds are failing because the branch is missing the latest changes from `master`. However, GHA packaging fails on setting the status, so the failure is not visible here.

File: developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFileAction.pas [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  Keyman.Developer.System.KmcWrapper;
+
+function TxmlLdmlProjectFileAction.Clean: Boolean;
+var
+  FJS: string;
+begin
+  CleanFile(OutputFileName);
+
+  if Targets * KMXKeymanTargets <> [] then
+    CleanFile(ExtractFilePath(FileName) + ExtractFileName(ChangeFileExt(KVKFileName, '.kvk')), True);
```
Comment (darcywong00) [inline developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFileAction.pas:47]: Why is KVKFileName used to establish the file name for CleanFile?

File: developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFile.pas [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+function TxmlLdmlProjectFile.GetTargetFilename: string;
+var
+  FTempFileVersion: string;
+begin
+  // https://github.com/keymanapp/keyman/issues/631
+  // This appears to be a Delphi compiler bug (RSP-20457)
+  // Workaround is to make a copy of the parameter locally
+  // which fixes the reference counting.
+  FTempFileVersion := FileVersion;
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFile.pas:136]: RSP-20457 is still not fixed.

File: developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+end;
+
+procedure TxmlLdmlProjectFileUI.SetDebug(const Value: Boolean);
+begin
+  ProjectFile.Debug := Value;
+end;
+
+function TxmlLdmlProjectFileUI.TestKeymanWeb(FSilent: Boolean): Boolean;   // I4409
+(*var
+  FCompiledName: string;
```
Comment (jahorton) [inline developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas:140]: This one of them multi-line comments of dead-code we usually remove?

File: developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        begin
+          FKeymanDeveloperOptions.AutoSaveBeforeCompiling := True;
+          FKeymanDeveloperOptions.Write;
+        end;
+      end;
+
+      if not modActionsMain.actFileSave.Execute then Exit;
+    end
+    else
+      Exit;
```
Comment (darcywong00) [inline developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas:107]: So when ProjecFile is modified and silent is true, this just exists as NOOP?

File: developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFile.pas [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    property KVKFileName: string read FKVKFileName;
+    property IsDebug: Boolean read FDebug;
+  public
+    class function IsFileTypeSupported(const Filename: string): Boolean; override;
+    function IsCompilable: Boolean; override;
+    procedure Load(node: IXMLNode); override;   // I4698
+    procedure Save(node: IXMLNode); override;   // I4698
+    procedure LoadState(node: IXMLNode); override;   // I4698
+    procedure SaveState(node: IXMLNode); override;   // I4698
```
Comment (jahorton) [inline developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFile.pas:44]: While I'm not quite following the inheritance and/or polymorphism patterns used here and in ProjectFile.pas (as noticed via commit 1), I can see a pretty strong correlation between the methods.

I'm mostly confused why these `procedure`s are `override` here but `virtual` there; I don't see the inheritance path between the two, and personally would've expected them to be sibling derived classes of a common base.  But those might be small nits and partly due to my inexperience with Delphi in particular.

File: developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas (hunk 3 above)
Comment (jahorton) [reply developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas:140]: Granted, I see this is the part where we'd test an _LDML_ keyboard in _Web_, which isn't going to be supported for this version.  But... surely just saying "Hey, just CTRL-F this function from this other class / file when the time comes" should work well enough?  Or do we expect significant differences in how it'd work?

File: developer/src/tike/actions/dmActionsDebugger.pas [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+    actDebugViewRegressionTesting: TAction;
+    actDebugTestMode: TAction;
+    actDebugDebuggerMode: TAction;
+    actDebugPause: TAction;
+    actDebugViewFont: TAction;
+    actDebugViewDefaultFont: TAction;
+    dlgFont: TFontDialog;
+    actDebugViewState: TAction;
+    actDebugSwitchToDebuggerWindow: TAction;
+    procedure actionsKeyboardEditorUpdate(Action: TBasicAction;
```
Comment (jahorton) [inline developer/src/tike/actions/dmActionsDebugger.pas:75]: As best as I can tell, the method above is new, but all that follow in this list have been migrated from `dmActionsKeyboardEditor` within this commit.

**Review by jahorton — APPROVED** [review]
For anything "actions", all I can give is an RSLGTM.  I'm not familiar with that abstraction from Core whatsoever, so I can't really give much feedback there.

For the rest of it, a plain ol' LGTM.

File: developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFileAction.pas (hunk 1 above)
Comment (mcdurdin) [reply developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFileAction.pas:47]: This is removing a build output, the .kvk file, if it exists.

File: developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas (hunk 3 above)
Comment (mcdurdin) [reply developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas:140]: This is on my list to address now in #9948.

File: developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFile.pas (hunk 5 above)
Comment (mcdurdin) [reply developer/src/tike/project/Keyman.Developer.System.Project.xmlLdmlProjectFile.pas:44]: The implementation in ProjectFile.pas is the base class, hence it is virtual. The other classes are the overrides and extend the base functionality.

File: developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas (hunk 3 above)
Comment (mcdurdin) [reply developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas:140]: (Oh yes, there will be significant differences in the kmn debugger and ldml debugger -- they will be completely divergent implementations, I expect)

File: developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas (hunk 4 above)
Comment (mcdurdin) [reply developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas:107]: Correct. It exits with a false return value so the caller knows that things didn't work out.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.211-alpha](https://keyman.com/downloads/releases/alpha/17.0.211)

**Pull request merged**